File: typescript/ec2-instance/src/libs/ec2Stack.ts
/* eslint-disable import/no-extraneous-dependencies */
import { Stack, StackProps, CfnOutput } from 'aws-cdk-lib';
import {  getCpuType, getInstanceClass, getInstanceSize, } from './modules/utils';
import { AmazonLinuxEdition, Instance, InstanceType, KeyPair, MachineImage, SecurityGroup, Vpc } from 'aws-cdk-lib/aws-ec2';
import { Construct } from 'constructs';

export interface Ec2Props extends StackProps {
  vpc: Vpc;
  sshSecurityGroup: SecurityGroup;
  cpuType: string;
  instanceSize: string;
  creator: string;
  count: string
}

export class Ec2Stack extends Stack {
  constructor(scope: Construct, id: string, props: Ec2Props) {
    super(scope, id, props);

    const keyPair = new KeyPair(this, `sshKey-${props.creator}`)

    for (let i = 1; i < Number(props.count)+1; i++) {
      // Create EC2 Instance
      const instance = new Instance(this, `Ec2Instance-${props.creator}${i}`, {
        vpc: props.vpc,
        machineImage: MachineImage.latestAmazonLinux2023({
          cachedInContext: false,
          cpuType: getCpuType(props.cpuType),
          edition: AmazonLinuxEdition.STANDARD,
        }),
        securityGroup: props.sshSecurityGroup,
        keyPair: keyPair,
        instanceType: InstanceType.of(
          getInstanceClass(props.cpuType),
          getInstanceSize(props.instanceSize)),
      });

      // SSH Command to connect to the EC2 Instance
      new CfnOutput(this, `sshCommand-${props.creator}${i}`, {
        value: `ssh ec2-user@${instance.instancePublicDnsName}`,
      });
    }

    new CfnOutput(this, `keyPair`, {
      value: `aws ssm get-parameter --name /ec2/keypair/${keyPair.keyPairId} --with-decryption --query Parameter.Value --output text`,
    });
  }
}

File: python/ec2-instance/libs/ec2Stack.py
from aws_cdk import (
    CfnOutput,
    Stack
)
from aws_cdk.aws_ec2 import (
    AmazonLinuxEdition,
    BlockDevice,
    BlockDeviceVolume,
    EbsDeviceProps,
    EbsDeviceVolumeType,
    Instance,
    InstanceType,
    MachineImage,
    KeyPair,
    SecurityGroup,
    Vpc
)
from constructs import Construct

from libs.modules.utils import (
    getCpuType,
    getInstanceClass,
    getInstanceSize
)

class Ec2Stack(Stack):
        
    def __init__(self, 
            scope: Construct,
            id: str,
            vpc:Vpc,
            sshSecurityGroup:SecurityGroup,
            cpuType:str,
            instanceSize:str,
            creator:str,
            count:str,
            **kwargs
        ) -> None:
        super().__init__(scope, id, **kwargs)
        
        keyPair = KeyPair(self,id=f"sshKey-{creator}")
        
        for i in range(1,int(count)+1):
            # Create EC2 Instance
            self.instance = Instance(self, f"Ec2Instance-{creator}{i}",
                vpc=vpc,
                machine_image=MachineImage.latest_amazon_linux2023(
                    cached_in_context=False,
                    cpu_type=getCpuType(cpuType),
                    edition=AmazonLinuxEdition.STANDARD
                ),
                security_group=sshSecurityGroup,
                key_pair=keyPair,
                instance_type=InstanceType.of(
                    instance_class=getInstanceClass(cpuType),
                    instance_size=getInstanceSize(instanceSize)
                )
            )
            
            # SSH Command to connect to the EC2 Instance
            CfnOutput(self, f"sshCommand-{creator}{i}",
                value=f'ssh ec2-user@{self.instance.instance_public_dns_name}'
            )
            
        CfnOutput(self, "keyPair",
                value=f'aws ssm get-parameter --name /ec2/keypair/${keyPair.key_pair_id} --with-decryption --query Parameter.Value --output text'
            )